tests/unit/sidecars/secrets_fetcher_tests.py

```diff
@@ -0,0 +1,220 @@
+import configparser
+import dataclasses
+import datetime
+import getpass
+import io
+import json
+import pathlib
+import sys
+import typing
+import unittest
+import unittest.mock
+
+from pyfakefs.fake_filesystem_unittest import TestCase
+
+from baseplate.lib import config
+from baseplate.sidecars import secrets_fetcher
+
+UTC: datetime.timezone
+if sys.version_info > (3, 11):
+    UTC = datetime.UTC
+else:
+    from datetime import timezone
+
+    UTC = timezone.utc
+
+whoami = getpass.getuser()
+
+configini = f"""
+[secret-fetcher]
+vault.url = https://vault.example.com:8200/
+vault.role = my-server-role
+vault.auth_type = aws
+vault.mount_point = aws-ec2
+
+output.path = /var/local/secrets.json
+output.owner = {getpass.getuser()}
+output.group = {getpass.getuser()}
+output.mode = 0400
+
+secrets =
+    secret/one,
+    secret/two,
+    secret/three,
+
+callback = scripts/my-transformer  # optional
+""".strip()
+
+
+@dataclasses.dataclass
+class FakeVaultClient(secrets_fetcher.VaultClient):
+    token_expiration: datetime.datetime
+    token: str = "token"
+
+    # @typing.override  # TODO: Added in version 3.12.
+    def get_secret(self, secret_name: str) -> tuple[typing.Any, datetime.datetime]:
+        return secret_name.upper(), self.token_expiration + datetime.timedelta(seconds=30)
+
+
+@dataclasses.dataclass
+class FakeBadVaultClient(secrets_fetcher.VaultClient):
+    token_expiration: datetime.datetime
+    token: str = "token-bad"
+
+    # @typing.override  # TODO: Added in version 3.12.
+    def get_secret(self, secret_name: str) -> tuple[typing.Any, datetime.datetime]:
+        """Return a secret value that is not JSON serializable."""
+        s = self.token_expiration
+        return s, self.token_expiration + datetime.timedelta(seconds=30)
+
+
+class Tests(TestCase):
+    @classmethod
+    def setUpClass(cls):
```

Comment:
Just curious, was there a reason to prefer unittest-style class tests rather than pytest style (test functions with fixtures) here? We have a mix of both so I don't think one is more correct than another, just wondering if there was anything motivating this choice over the other.

Reply:
Nothing except that I was actually able to get this working in this style. There is a [pytest plugin](https://pytest-pyfakefs.readthedocs.io/en/stable/usage.html#patch-using-the-pytest-plugin).